File: packages/intutic-clawde/intutic_clawde/control_plane.py
"""Control-plane management client (LLD #69) -- org/team/gateway/credentials
administration, as distinct from ClawdeClient's data-plane chat calls.

Deliberately a separate class, not new methods on ClawdeClient:
ClawdeClient.base_url targets the *proxy* (default http://localhost:4000);
the control plane is a different origin entirely (default
https://app.intutic.ai, or a self-hosted CONTROL_PLANE_URL). Bolting
management calls onto ClawdeClient would silently need a second base URL
on a class whose whole contract today is "one client, one proxy."

Every endpoint here is mirrored 1:1 from tools/cli/src/commands/{org,team,
gateway,credentials,whoami}.ts -- the CLI's own already-tested contracts,
not re-derived. Endpoints deliberately NOT included: session establishment
(login/logout -- an SDK caller supplies api_key directly, matching
ClawdeClient's own existing constructor contract), and local-environment/
terminal concerns (init, doctor, install-daemon, integrity, rollback,
connect, exec, start, sync-context, skill) that have no meaning for a
library embedded in someone else's process.

Works unmodified against a self-hosted control plane -- there is no
SaaS-vs-self-hosted branch anywhere in this file. An open-core user who
runs the proxy standalone with no control plane configured simply never
constructs this class (or does, points it at nothing, and gets a
ClawdeConnectionError on the first call) -- the same framing whoami.ts
already uses: "This command needs an Intutic control plane, which open
core does not include."

Auth: the same api_key type ClawdeClient accepts -- a vk_ token or a JWT
both satisfy services/control-plane/src/middleware/auth.ts, which resolves
either to an AuthContext with a role; endpoints gated
requireRole('OWNER','ADMIN') server-side reject an under-privileged token
exactly as they would for the CLI using the same token.
"""

import os
import requests
from typing import Any, Dict, List, Optional
from .errors import ClawdeConnectionError

DEFAULT_CONTROL_PLANE_URL = "https://app.intutic.ai"


class ControlPlaneClient:
    def __init__(self, api_key: str, base_url: Optional[str] = None):
        if not api_key:
            raise ValueError("API key is required to initialize ControlPlaneClient.")
        self.api_key = api_key
        self.base_url = (
            base_url
            or os.environ.get("INTUTIC_CONTROL_PLANE_URL")
            or DEFAULT_CONTROL_PLANE_URL
        )

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        auth: bool = True,
    ) -> Any:
        headers: Dict[str, str] = {"Content-Type": "application/json"}
        if auth:
            headers["Authorization"] = f"Bearer {self.api_key}"

        try:
            res = requests.request(
                method, f"{self.base_url}{path}", json=body, headers=headers, timeout=30.0
            )
        except requests.RequestException as e:
            raise ClawdeConnectionError(
                f"Could not reach control plane at {self.base_url}: {e}"
            )

        if not res.ok:
            raise ClawdeConnectionError(
                f"Control plane {method} {path} failed ({res.status_code}): {res.text}"
            )

        if not res.content:
            return None
        return res.json()

    def whoami(self) -> Dict[str, Any]:
        """GET /api/v1/auth/me"""
        return self._request("GET", "/api/v1/auth/me")

    def signup_org(self, email: str, password: str, name: str, org_name: str) -> Dict[str, Any]:
        """POST /api/v1/auth/signup/org -- unauthenticated, creates the calling user."""
        return self._request(
            "POST",
            "/api/v1/auth/signup/org",
            {"email": email, "password": password, "name": name, "orgName": org_name},
            auth=False,
        )

    def list_teams(self, org_id: str) -> List[Dict[str, Any]]:
        """GET /api/v1/orgs/:orgId/teams"""
        res = self._request("GET", f"/api/v1/orgs/{_quote(org_id)}/teams")
        return res.get("data", []) if res else []

    def create_team(self, org_id: str, name: str) -> Dict[str, Any]:
        """POST /api/v1/orgs/:orgId/teams"""
        return self._request("POST", f"/api/v1/orgs/{_quote(org_id)}/teams", {"name": name})

    def list_team_workspaces(self, team_id: str) -> List[Dict[str, Any]]:
        """GET /api/v1/teams/:teamId/workspaces"""
        res = self._request("GET", f"/api/v1/teams/{_quote(team_id)}/workspaces")
        return res.get("data", []) if res else []

    def create_workspace(self, team_id: str, name: str) -> Dict[str, Any]:
        """POST /api/v1/teams/:teamId/workspaces"""
        return self._request("POST", f"/api/v1/teams/{_quote(team_id)}/workspaces", {"name": name})

    def register_gateway(self, name: str, deployment_target: str) -> Dict[str, Any]:
        """POST /api/v1/gateways"""
        return self._request(
            "POST", "/api/v1/gateways", {"name": name, "deploymentTarget": deployment_target}
        )

    def list_gateways(self) -> List[Dict[str, Any]]:
        """GET /api/v1/gateways"""
        res = self._request("GET", "/api/v1/gateways")
        return res.get("data", []) if res else []

    def get_gateway_status(self, gateway_id: str) -> Dict[str, Any]:
        """GET /api/v1/gateways/:id/status"""
        return self._request("GET", f"/api/v1/gateways/{_quote(gateway_id)}/status")

    def rotate_gateway_token(self, gateway_id: str) -> Dict[str, Any]:
        """POST /api/v1/gateways/:id/rotate"""
        return self._request("POST", f"/api/v1/gateways/{_quote(gateway_id)}/rotate", {})

    def revoke_gateway(self, gateway_id: str, reason: Optional[str] = None) -> None:
        """DELETE /api/v1/gateways/:id"""
        self._request("DELETE", f"/api/v1/gateways/{_quote(gateway_id)}", {"reason": reason})

    def set_gateway_config(
        self,
        gateway_id: str,
        require_vk: Optional[bool] = None,
        require_provisioned_key: Optional[bool] = None,
    ) -> Dict[str, Any]:
        """PATCH /api/v1/gateways/:id/config"""
        body: Dict[str, Any] = {}
        if require_vk is not None:
            body["requireVk"] = require_vk
        if require_provisioned_key is not None:
            body["requireProvisionedKey"] = require_provisioned_key
        return self._request("PATCH", f"/api/v1/gateways/{_quote(gateway_id)}/config", body)

    def assign_workspace_gateway(self, gateway_id: Optional[str]) -> Dict[str, Any]:
        """PATCH /api/v1/workspace/gateway -- pass None to clear the override."""
        return self._request("PATCH", "/api/v1/workspace/gateway", {"gatewayId": gateway_id})

    def assign_org_gateway(self, org_id: str, gateway_id: Optional[str]) -> Dict[str, Any]:
        """PATCH /api/v1/orgs/:orgId/gateway -- pass None to clear the org default."""
        return self._request(
            "PATCH", f"/api/v1/orgs/{_quote(org_id)}/gateway", {"gatewayId": gateway_id}
        )

    def resolve_gateway(self) -> Dict[str, Any]:
        """GET /api/v1/workspace/gateway-resolution"""
        return self._request("GET", "/api/v1/workspace/gateway-resolution")

    def list_provider_credentials(self) -> List[Dict[str, Any]]:
        """GET /api/v1/workspace/provider-credentials"""
        res = self._request("GET", "/api/v1/workspace/provider-credentials")
        return res.get("data", []) if res else []

    def set_provider_credential(self, provider: str, fields: Dict[str, str]) -> Dict[str, Any]:
        """PUT /api/v1/workspace/provider-credentials/:provider"""
        return self._request(
            "PUT", f"/api/v1/workspace/provider-credentials/{_quote(provider)}", fields
        )

    def unset_provider_credential(self, provider: str) -> None:
        """DELETE /api/v1/workspace/provider-credentials/:provider"""
        self._request("DELETE", f"/api/v1/workspace/provider-credentials/{_quote(provider)}")


def _quote(segment: str) -> str:
    from urllib.parse import quote

    return quote(segment, safe="")

File: packages/clawde-sdk/src/control-plane.ts
import { ClawdeConnectionError } from './errors'
import type {
  ControlPlaneClientOptions,
  WhoamiResult,
  OrgSignupParams,
  OrgSignupResult,
  StartDomainVerificationResult,
  CheckDomainVerificationResult,
  CreateOrgParams,
  CreateOrgResult,
  Team,
  Workspace,
  GatewayRegisterParams,
  GatewayRegisterResult,
  Gateway,
  GatewayRotateResult,
  GatewayStatus,
  GatewayConfigUpdate,
  GatewayConfigResult,
  GatewayResolution,
  ProviderCredentialStatus,
} from './types'

/**
 * Control-plane management client (LLD #69) — org/team/gateway/credentials
 * administration, as distinct from `ClawdeClient`'s data-plane chat calls.
 *
 * Deliberately a separate class, not new methods on `ClawdeClient`:
 * `ClawdeClient.baseUrl` targets the *proxy* (default `http://localhost:4000`);
 * the control plane is a different origin entirely (default
 * `https://app.intutic.ai`, or a self-hosted `CONTROL_PLANE_URL`). Bolting
 * management calls onto `ClawdeClient` would silently need a second base URL
 * on a class whose whole contract today is "one client, one proxy."
 *
 * Every endpoint here is mirrored 1:1 from `tools/cli/src/commands/{org,
 * team,gateway,credentials,whoami}.ts` — the CLI's own already-tested
 * contracts, not re-derived. Endpoints deliberately NOT included: session
 * establishment (`login`/`logout` — an SDK caller supplies `apiKey`
 * directly), and local-environment/terminal concerns (`init`, `doctor`,
 * `install-daemon`, `integrity`, `rollback`, `connect`, `exec`, `start`,
 * `syncContext`, `skill`) that have no meaning for a library embedded in
 * someone else's process.
 *
 * Works unmodified against a self-hosted control plane — there is no
 * SaaS-vs-self-hosted branch anywhere in this file. An open-core user who
 * runs the proxy standalone with no control plane configured simply never
 * constructs this class (or does, points it at nothing, and gets a
 * `ClawdeConnectionError` on the first call) — the same framing
 * `whoami.ts` already uses: "This command needs an Intutic control plane,
 * which open core does not include."
 *
 * Auth: the same `apiKey` type `ClawdeClient` accepts — a `vk_` token or a
 * JWT both satisfy `services/control-plane/src/middleware/auth.ts`, which
 * resolves either to an `AuthContext` with a role; endpoints gated
 * `requireRole('OWNER','ADMIN')` server-side reject an under-privileged
 * token exactly as they would for the CLI using the same token.
 */
export class ControlPlaneClient {
  private apiKey: string
  private baseUrl: string

  constructor(options: ControlPlaneClientOptions) {
    if (!options.apiKey) {
      throw new Error('API key is required to initialize ControlPlaneClient.')
    }
    this.apiKey = options.apiKey
    this.baseUrl =
      options.baseUrl || process.env.INTUTIC_CONTROL_PLANE_URL || 'https://app.intutic.ai'
  }

  private async request<T>(method: string, path: string, body?: unknown, auth = true): Promise<T> {
    const headers: Record<string, string> = { 'Content-Type': 'application/json' }
    if (auth) headers['Authorization'] = `Bearer ${this.apiKey}`

    let res: Response
    try {
      res = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
      })
    } catch (err: any) {
      throw new ClawdeConnectionError(`Could not reach control plane at ${this.baseUrl}: ${err.message}`)
    }

    if (!res.ok) {
      const text = await res.text().catch(() => 'Unknown error')
      throw new ClawdeConnectionError(`Control plane ${method} ${path} failed (${res.status}): ${text}`)
    }

    return (await res.json()) as T
  }

  /** GET /api/v1/auth/me */
  public async whoami(): Promise<WhoamiResult> {
    return this.request('GET', '/api/v1/auth/me')
  }

  /**
   * POST /api/v1/auth/signup/org — unauthenticated, creates the calling user.
   *
   * Closed by default in production (`INTUTIC_PUBLIC_ORG_SIGNUP`, off unless
   * a deployment has built its own anonymous domain-verification story):
   * creating a real org auto-provisions a real managed gateway cell (LLD
   * #71), so org creation now requires DNS domain-ownership proof, and an
   * anonymous caller has no session to own a verification attempt against.
   * Prefer `startDomainVerification` + `checkDomainVerification` +
   * `createOrg` with an already-authenticated `apiKey` — the same flow the
   * CLI's `intutic org create` and the dashboard's "Create Organization"
   * modal use.
   */
  public async signupOrg(params: OrgSignupParams): Promise<OrgSignupResult> {
    return this.request('POST', '/api/v1/auth/signup/org', params, false)
  }

  /**
   * POST /api/v1/domain-verification/start — mints a DNS TXT-record
   * verification token for `domain`. Publish a TXT record at
   * `txtRecordName` with value `txtRecordValue`, then poll
   * `checkDomainVerification` until `status` is `'verified'`.
   */
  public async startDomainVerification(domain: string): Promise<StartDomainVerificationResult> {
    return this.request('POST', '/api/v1/domain-verification/start', { domain })
  }

  /**
   * GET /api/v1/domain-verification/:id — re-checks DNS for the TXT record.
   * Safe to call repeatedly; performs a fresh lookup every call.
   */
  public async checkDomainVerification(verificationId: string): Promise<CheckDomainVerificationResult> {
    return this.request('GET', `/api/v1/domain-verification/${encodeURIComponent(verificationId)}`)
  }

  /**
   * POST /api/v1/orgs — creates a real org from an already-authenticated
   * caller. Requires a `verified`, unconsumed domain verification (see
   * `startDomainVerification`); the verification is consumed atomically
   * inside the org-insert transaction, so it backs exactly this one org.
   */
  public async createOrg(params: CreateOrgParams): Promise<CreateOrgResult> {
    return this.request('POST', '/api/v1/orgs', params)
  }

  /** GET /api/v1/orgs/:orgId/teams */
  public async listTeams(orgId: string): Promise<Team[]> {
    const res = await this.request<{ data: Team[] }>('GET', `/api/v1/orgs/${encodeURIComponent(orgId)}/teams`)
    return res.data ?? []
  }

  /** POST /api/v1/orgs/:orgId/teams */
  public async createTeam(orgId: string, name: string): Promise<Team> {
    return this.request('POST', `/api/v1/orgs/${encodeURIComponent(orgId)}/teams`, { name })
  }

  /** GET /api/v1/teams/:teamId/workspaces */
  public async listTeamWorkspaces(teamId: string): Promise<Workspace[]> {
    const res = await this.request<{ data: Workspace[] }>(
      'GET',
      `/api/v1/teams/${encodeURIComponent(teamId)}/workspaces`,
    )
    return res.data ?? []
  }

  /** POST /api/v1/teams/:teamId/workspaces */
  public async createWorkspace(teamId: string, name: string): Promise<Workspace> {
    return this.request('POST', `/api/v1/teams/${encodeURIComponent(teamId)}/workspaces`, { name })
  }

  /** POST /api/v1/gateways */
  public async registerGateway(params: GatewayRegisterParams): Promise<GatewayRegisterResult> {
    return this.request('POST', '/api/v1/gateways', params)
  }

  /** GET /api/v1/gateways */
  public async listGateways(): Promise<Gateway[]> {
    const res = await this.request<{ data: Gateway[] }>('GET', '/api/v1/gateways')
    return res.data ?? []
  }

  /** GET /api/v1/gateways/:id/status */
  public async getGatewayStatus(gatewayId: string): Promise<GatewayStatus> {
    return this.request('GET', `/api/v1/gateways/${encodeURIComponent(gatewayId)}/status`)
  }

  /** POST /api/v1/gateways/:id/rotate */
  public async rotateGatewayToken(gatewayId: string): Promise<GatewayRotateResult> {
    return this.request('POST', `/api/v1/gateways/${encodeURIComponent(gatewayId)}/rotate`, {})
  }

  /** DELETE /api/v1/gateways/:id */
  public async revokeGateway(gatewayId: string, reason?: string): Promise<void> {
    await this.request('DELETE', `/api/v1/gateways/${encodeURIComponent(gatewayId)}`, { reason })
  }

  /** PATCH /api/v1/gateways/:id/config */
  public async setGatewayConfig(gatewayId: string, config: GatewayConfigUpdate): Promise<GatewayConfigResult> {
    return this.request('PATCH', `/api/v1/gateways/${encodeURIComponent(gatewayId)}/config`, config)
  }

  /** PATCH /api/v1/workspace/gateway — pass null to clear the override. */
  public async assignWorkspaceGateway(gatewayId: string | null): Promise<{ gatewayId: string | null }> {
    return this.request('PATCH', '/api/v1/workspace/gateway', { gatewayId })
  }

  /** PATCH /api/v1/orgs/:orgId/gateway — pass null to clear the org default. */
  public async assignOrgGateway(orgId: string, gatewayId: string | null): Promise<{ gatewayId: string | null }> {
    return this.request('PATCH', `/api/v1/orgs/${encodeURIComponent(orgId)}/gateway`, { gatewayId })
  }

  /** GET /api/v1/workspace/gateway-resolution */
  public async resolveGateway(): Promise<GatewayResolution> {
    return this.request('GET', '/api/v1/workspace/gateway-resolution')
  }

  /** GET /api/v1/workspace/provider-credentials */
  public async listProviderCredentials(): Promise<ProviderCredentialStatus[]> {
    const res = await this.request<{ data: ProviderCredentialStatus[] }>(
      'GET',
      '/api/v1/workspace/provider-credentials',
    )
    return res.data ?? []
  }

  /** PUT /api/v1/workspace/provider-credentials/:provider */
  public async setProviderCredential(
    provider: string,
    fields: Record<string, string>,
  ): Promise<ProviderCredentialStatus> {
    return this.request('PUT', `/api/v1/workspace/provider-credentials/${encodeURIComponent(provider)}`, fields)
  }

  /** DELETE /api/v1/workspace/provider-credentials/:provider */
  public async unsetProviderCredential(provider: string): Promise<void> {
    await this.request('DELETE', `/api/v1/workspace/provider-credentials/${encodeURIComponent(provider)}`)
  }
}
